kafka-streams-partitioners/weighted-group-partitioner/src/main/java/org/hypertrace/core/kafkastreams/framework/partitioner/WeightedGroupPartitioner.java

```diff
@@ -0,0 +1,69 @@
+package org.hypertrace.core.kafkastreams.framework.partitioner;
+
+import com.google.common.base.Preconditions;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import javax.annotation.Nonnull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.streams.processor.StreamPartitioner;
+import org.hypertrace.core.kafkastreams.framework.partitioner.WeightedGroupProfile.WeightedGroup;
+
+@Slf4j
+public class WeightedGroupPartitioner<K, V> implements StreamPartitioner<K, V> {
+  private final String profileName;
+  private final PartitionerConfigServiceClient configServiceClient;
+  private final BiFunction<K, V, String> memberIdExtractor;
+  private final StreamPartitioner<K, V> delegatePartitioner;
+
+  public WeightedGroupPartitioner(
+      @Nonnull PartitionerConfigServiceClient configServiceClient,
+      @Nonnull String profileName,
+      @Nonnull BiFunction<K, V, String> memberIdExtractor,
+      @Nonnull StreamPartitioner<K, V> delegatePartitioner) {
+
+    Preconditions.checkNotNull(configServiceClient);
+    Preconditions.checkNotNull(profileName);
+    Preconditions.checkNotNull(memberIdExtractor);
+    Preconditions.checkNotNull(delegatePartitioner);
+
+    this.configServiceClient = configServiceClient;
+    this.profileName = profileName;
+    this.memberIdExtractor = memberIdExtractor;
+    this.delegatePartitioner = delegatePartitioner;
+  }
+
+  @Override
+  public Integer partition(String topic, K key, V value, int numPartitions) {
+    String memberId = getMemberId(topic, key, value);
+
+    WeightedGroup groupConfig = this.getPartitionGroupInfo(memberId);
+    int fromIndex = (int) Math.floor(groupConfig.getNormalizedFractionalStart() * numPartitions);
+    int toIndex = (int) Math.ceil(groupConfig.getNormalizedFractionalEnd() * numPartitions);
+    int numPartitionsForGroup = toIndex - fromIndex;
+
+    // partitioner by contract can return null.
+    // Refer api doc:  org.apache.kafka.streams.processor.StreamPartitioner.partition
+    // when delegate partitioner returns null, we treat it as 0
```

Comment:
not sure this is right - a delegate partitioner that returns null often (which is an indication for the caller to use default partitioning) would result in a skew in that first partition. We'd probably want to use an `orElse(roundRobin)` style behavior - some evenly distributed option that promises to never return null.

Reply:
fixed. used round robin when delegate returns null